### PR TITLE
Feature/rating backend

### PR DESCRIPTION
**Pull request by ElBatanony**
Use Firebase Cloud Functions to update the ratings in the ta-course documents.
Calculate and display the average rating (client-side).

### CODE REVIEW

File: firebase/functions/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import * as functions from 'firebase-functions';
+import * as admin from 'firebase-admin';
+
+admin.initializeApp();
+
+const db = admin.firestore();
+
+exports.modifyRating = functions.firestore
+    .document('ta-course/{taCourseId}/ratings/{uid}')
```
Comment (Flamel001) [inline firebase/functions/src/index.ts:9]: how does it work, where do these taCourseId and uid came from?

File: firebase/functions/package.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  "engines": {
+    "node": "12"
+  },
+  "main": "lib/index.js",
+  "dependencies": {
+    "firebase-admin": "^9.2.0",
+    "firebase-functions": "^3.11.0"
+  },
+  "devDependencies": {
+    "typescript": "^3.8.0",
```
Comment (Flamel001) [inline firebase/functions/package.json:20]: mMm tasty

File: firebase/functions/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+        const newData = change.after.exists ? change.after.data() : null;
+        const oldData = change.before.exists ? change.before.data() : null;
+
+        const taCourseDoc = db.collection('ta-course').doc(taCourseId)
+
+        let newTaCourse: any = {};
+
+        if (newData && newData.rating)
+            newTaCourse['rating' + newData.rating.toString()] = admin.firestore.FieldValue.arrayUnion(uid);
```
Comment (Flamel001) [inline firebase/functions/src/index.ts:23]: why and object prop could have value 'rating5'?) Or could you explain how does this work?

File: lib/data.dart [hunk 4]
```diff
@@ -22,27 +22,35 @@ class Student {
...
+  double avgRating;
+  List<int> ratings;
+
+  calculateAvgRating() {
+    int ratingsCount = ratings.reduce((value, element) => value + element);
+    int sumRatings = 0;
+    for (int i = 0; i < 5; i += 1) {
+      sumRatings += (i + 1) * ratings[i];
+    }
+    avgRating = (sumRatings / (ratingsCount * 5)) * 5;
```
Comment (Flamel001) [inline lib/data.dart]: `avgRating = ratingCount/ratings.length` or I don't understand what do you want🤔 And I believe some return statement is required

**Review by Flamel001 — CHANGES_REQUESTED** [review]

File: firebase/functions/src/index.ts (hunk 1 above)
Comment (ElBatanony) [reply firebase/functions/src/index.ts:9]: This is a wildcard. You can find more info [here](https://firebase.google.com/docs/functions/firestore-events#wildcards-parameters).

File: firebase/functions/src/index.ts (hunk 3 above)
Comment (ElBatanony) [reply firebase/functions/src/index.ts:23]: I think it is better to show. This is how the doc looks like. Each rating (1-5) is a list of user ids of users who provided the rating.
image

File: lib/data.dart (hunk 4 above)
Comment (ElBatanony) [reply lib/data.dart]: No return is required. It is saved as a class member.
I will double-check the math.

Comment (ElBatanony) [reply lib/data.dart]: Fixed.

**Pull request merged**